### PR TITLE
iOS `presentationStyle` Modal Appearance

### PR DESCRIPTION
**Pull request by hnryjms**
#2613 - Added `presentationStyle` for larger iOS devices modal appearance

## Motivation

When using `<Modal` on larger iOS devices, esp. iPad and iPhone 7 Plus devices, there is no way to use the system functionality for controlling the appearance of modals (`presentationStyle`), which improves the native system's animation and display of smaller content appearing within large horizontal space.

## Test Plan

I've added a new picker for selecting a `presentationStyle` within  the RNTester app. See below for the appearance of this change, as well as the relevant changes to the RN documentation.

may-22-2017 09-49-50
screen shot 2017-05-22 at 9 50 12 am

### CODE REVIEW

File: React/Views/RCTModalHostView.h [hunk 1]
```diff
@@ -21,6 +21,7 @@
 @interface RCTModalHostView : UIView <RCTInvalidating>
 
 @property (nonatomic, copy) NSString *animationType;
+@property (nonatomic, copy) NSString *presentationStyle;
```
Comment (shergin) [inline React/Views/RCTModalHostView.h]: Please use `RCTConvert` to map strings to enum values (see `RCTScrollView` for examples).

File: React/Views/RCTModalHostView.m [hunk 2]
```diff
@@ -165,7 +176,9 @@ - (BOOL)hasAnimationType
 
 - (void)setTransparent:(BOOL)transparent
 {
-  _modalViewController.modalPresentationStyle = transparent ? UIModalPresentationOverFullScreen : UIModalPresentationFullScreen;
+  if ([self isTransparent] != transparent) {
```
Comment (shergin) [inline React/Views/RCTModalHostView.m]: Dot notation here.

File: React/Views/RCTModalHostViewManager.m [hunk 3]
```diff
@@ -91,6 +91,7 @@ - (void)invalidate
 }
 
 RCT_EXPORT_VIEW_PROPERTY(animationType, NSString)
+RCT_EXPORT_VIEW_PROPERTY(presentationStyle, NSString)
```
Comment (shergin) [inline React/Views/RCTModalHostViewManager.m]: `RCT_EXPORT_VIEW_PROPERTY(presentationStyle, UIModalPresentationStyle)`

File: Libraries/Modal/Modal.js [hunk 4]
```diff
@@ -138,6 +153,18 @@ class Modal extends React.Component {
   static contextTypes = {
     rootTag: PropTypes.number,
   };
+  constructor(props: Object) {
+    super(props);
+    Modal._confirmProps(props);
+  }
+  componentWillReceiveProps(nextProps: Object) {
```
Comment (shergin) [inline Libraries/Modal/Modal.js]: New line.

File: Libraries/Modal/Modal.js [hunk 5]
```diff
@@ -138,6 +153,18 @@ class Modal extends React.Component {
...
+    Modal._confirmProps(nextProps);
+  }
+  static _confirmProps(props: Object) {
```
Comment (shergin) [inline Libraries/Modal/Modal.js]: New line.

File: Libraries/Modal/Modal.js (hunk 4 above)
Comment (hnryjms) [inline Libraries/Modal/Modal.js]: Could you send me a link to a code style spec document? I had been assuming that `npm run lint` and @eslint-bot had been the declarative code style requirements. If not, a link would be helpful both now, and committed in the `CONTRIBUTING.md` file.

File: React/Views/RCTModalHostView.m [hunk 6]
```diff
@@ -176,7 +168,7 @@ - (BOOL)hasAnimationType
 
 - (void)setTransparent:(BOOL)transparent
 {
-  if ([self isTransparent] != transparent) {
+  if (self.transparent != transparent) {
```
Comment (shergin) [inline React/Views/RCTModalHostView.m]: `self.isTransparent`, no?
I am curious, how does it work? We rename the getter to `isTransparent`, but... accessing it as `self.transparent` still works. That's confusing.

File: Libraries/Modal/Modal.js [hunk 7]
```diff
@@ -153,16 +154,19 @@ class Modal extends React.Component {
...
   }
+
   componentWillReceiveProps(nextProps: Object) {
     Modal._confirmProps(nextProps);
   }
+
   static _confirmProps(props: Object) {
     if (props.presentationStyle && props.presentationStyle !== 'overFullScreen' && props.transparent) {
-      console.warn(`Modal with '${props.presentationStyle}' presentation style and 'transparent' value is not supported.`);
+      warning(false, `Modal with '${props.presentationStyle}' presentation style and 'transparent' value is not supported.`);
```
Comment (shergin) [inline Libraries/Modal/Modal.js]: You have to put condition inside `warning`, that's the point of `warning`.

File: React/Views/RCTModalHostView.m (hunk 6 above)
Comment (hnryjms) [inline React/Views/RCTModalHostView.m]: That's ObjC, man 😶

File: Libraries/Modal/Modal.js (hunk 7 above)
Comment (hnryjms) [inline Libraries/Modal/Modal.js]: Seems to me that throughout the code, using `warning(false, ...)` is preferable. See links below:

https://github.com/facebook/react-native/blob/e13b9c6e49480e8262df06b7c1e99caab74e801f/Libraries/Components/TextInput/TextInput.js#L644

https://github.com/facebook/react-native/blob/5257c35d05578bab0233254da81f1216f461f18d/Libraries/Utilities/BackAndroid.js#L35

https://github.com/facebook/react-native/blob/9b4a644fec94422c2976f11be0ccf66f2715d9bd/Libraries/Network/XMLHttpRequest.js#L185

Comment (shergin) [inline Libraries/Modal/Modal.js]: Yeah, but it should be something like

`warning(!(props.presentationStyle && props.presentationStyle !== 'overFullScreen' && props.transparent), ...)`
or
`warning(!props.presentationStyle || props.presentationStyle === 'overFullScreen' || !props.transparent, ...)`
.

File: React/Views/RCTModalHostView.m (hunk 6 above)
Comment (shergin) [inline React/Views/RCTModalHostView.m]: Yeah, I know, but how? I prefer to see `self.isTransparent` here anyway.

Comment (hnryjms) [inline React/Views/RCTModalHostView.m]: Generally ObjC prefers using bracket syntax for synthesized getters, or dot syntax for property reference directly. This is the most common practice, so I am going to leave it as is.

**Comment by shergin** [comment]
@hnryjms Thank you! I will merge it on Monday.

**Comment by hnryjms** [comment]
Any update on this?

**Comment by hnryjms** [comment]
🎉

File: React/Views/RCTModalHostView.m [hunk 8]
```diff
@@ -168,7 +168,7 @@ - (BOOL)hasAnimationType
 
 - (void)setTransparent:(BOOL)transparent
 {
-  if (_transparent == transparent) {
+  if (self.isTransparent != transparent) {
```
Comment (hnryjms) [inline React/Views/RCTModalHostView.m]: Weird...your change recommended earlier shouldn't have affected CI systems....but it must have messed up the internal Facebook tools (which I don't know).... Guess I was lucky 🤷‍♂️

File: Libraries/Modal/Modal.js [hunk 9]
```diff
@@ -138,6 +153,18 @@ class Modal extends React.Component {
...
+  constructor(props: Object) {
+    super(props);
+    Modal._confirmProps(props);
+  }
+  componentWillReceiveProps(nextProps: Object) {
+    Modal._confirmProps(nextProps);
+  }
+  static _confirmProps(props: Object) {
+    if (props.presentationStyle && props.presentationStyle !== 'overFullScreen' && props.transparent) {
+      console.warn(`Modal with '${props.presentationStyle}' presentation style and 'transparent' value is not supported.`);
```
Comment (luco) [inline Libraries/Modal/Modal.js:168]: @shergin I couldn't find where in the Swift/Obj-C documentation it says using presentationStyle with transparent it's invalid.

https://developer.apple.com/documentation/uikit/uimodalpresentationstyle/formsheet

**Pull request closed without merging**